r/R/record-batch.R

```diff
@@ -189,3 +189,59 @@ record_batch <- RecordBatch$create
 
 #' @export
 names.RecordBatch <- function(x) x$names()
+
+#' @export
+rbind.RecordBatch <- function(...) {
+  abort("Use `Table$create()` to combine record batches")
+}
+
+cbind_check_length <- function(inputs, call = caller_env()) {
+  sizes <- map(inputs, function(input) {
+    # TODO: can we somehow implement vec_size for Arrow structures?
+    if (inherits(input, "ArrowTabular") || is.data.frame(input)) {
+      nrow(input)
+    } else {
+      length(input)
+    }
+  })
+  wrong_idx <- which.min(sizes == sizes[[1]] | sizes == 1)
+  if (wrong_idx != 1) {
+    abort(
+      c("Non-scalar inputs must have an equal number of rows.",
+        i = sprintf("..1 has %d, ..%d has %d", sizes[[1]], wrong_idx, sizes[[wrong_idx]])),
+      call = call
+    )
+  }
+}
+
+#' @export
+cbind.RecordBatch <- function(...) {
+  call <- sys.call()
+  inputs <- list(...)
+  arg_names <- if (is.null(names(inputs))) {
+    rep("", length(inputs))
+  } else {
+    names(inputs)
+  }
+
+  cbind_check_length(inputs, call)
+
+  columns <- flatten(map(seq_along(inputs), function(i) {
+    input <- inputs[[i]]
+    name <- arg_names[i]
+
+    if (inherits(input, "RecordBatch")) {
+      set_names(input$columns, names(input))
+    } else if (inherits(input, "data.frame")) {
+      as.list(input)
+    } else {
+      if (is.na(name) || name == "") {
```

Comment:
You know I'm not really sure why I handle NA here. Checking for `""` is useful because that's the behavior if only *some* of the args are missing:

```r
> test <- function(...) { names(list(...)) }
> test(x = 2, 4)
[1] "x" ""
> test(2, 4)
NULL
```